File: 2014/s3/s3.py
#Open input file
File = open("s3.in",'r')

#Read the number of tests to be run
T = int(File.readline())

#Run the number of tests
for i in range(T):

	#Read the number of carts in the test
	N = int(File.readline())

	#List containing the carts
	carts = []
	for j in range(N):
		carts.append(int(File.readline()))
	#List containing carts on branch
	branch = [0]
	#List containing carts at lake
	lake = [0]

	#The state flag
	state = "Y"

	#Run the test while the state is "Y" and all the carts haven't arrived at the lake
	while state == "Y" and lake[len(lake)-1] < N:
		if not len(carts) == 0 and carts[len(carts)-1] == lake[len(lake)-1] + 1:
			lake.append(carts.pop())
		elif branch[len(branch)-1] == lake[len(lake)-1] + 1:
			lake.append(branch.pop())
		elif not len(carts) == 0:
			branch.append(carts.pop())
		else:
			state = "N"
	print state

File: 2014/s3/s3.cpp
#include <stdio.h>
#include <stack>

using namespace std;

int main(int argc, char *argv[])
{
	//Read the number of tests
	int T = 0;
	scanf("%d",&T);

	//Run the tests
	for(int i = 0; i < T; ++i)
	{
		//The carts on the mountain
		stack<int> mountain;
		mountain.push(0);
		//The carts on the branch
		stack<int> branch;
		branch.push(0);
		//The top carts in the lake
		int lake = 0;
		
		//Read the number of carts
		int N = 0;
		scanf("%d",&N);


		//Read the carts
		for(int j = 0; j < N; ++j)
		{
			int cart = 0;
			scanf("%d",&cart);
			mountain.push(cart);
		}

		// 'Y' if all carts can make it to the lake 'N' if not
		char state = 'Y';

		//While the puzzle is still possible and has not been completed
		while(state == 'Y' && lake != N)
		{
			//If the cart on the mountain can be moved to the lake, move it
			if(mountain.top() == lake + 1)
			{
				lake++;
				mountain.pop();
			}
			//If the carts on the branch can be moved to the lake move it
			else if(branch.top() == lake + 1)
			{
				lake++;
				branch.pop();
			}
			//If there is a cart on the mountain move it to the branch
			else if(mountain.size() > 1)
			{
				branch.push(mountain.top());
				mountain.pop();
			}
			//Otherwise it is impossible
			else
			{
				state = 'N';
			}
		}

		printf("%c\n",state);
	}
	return 0;
}
